### PR TITLE
Granular invalidateblock and RewindBlockIndex

### PR DESCRIPTION
**Pull request by sipa**
This PR makes a number of improvements to the InvalidateBlock (`invalidateblock` RPC) and RewindBlockIndex functions, primarily around breaking up their long-term cs_main holding. In addition:
* They're made safely interruptible (`bitcoind` can be shutdown, and no progress in either will be lost, though if incomplete, `invalidateblock` won't continue after restart and will need to be called again)
* The validation queue is prevented from overflowing (meaning `invalidateblock` on a very old block will not drive bitcoind OOM) (see #14289).
* `invalidateblock` won't bother to move transactions back into the mempool after 10 blocks (optimization).

This is not an optimal solution, as we're relying on the scheduler call sites to make sure the scheduler doesn't overflow. Ideally, the scheduler would guarantee this directly, but that needs a few further changes (moving the signal emissions out of cs_main) to prevent deadlocks.

I have manually tested the `invalidateblock` changes (including interrupting, and running with -checkblockindex and -checkmempool), but haven't tried the rewinding (which is probably becoming increasingly unnecessary, as very few pre-0.13.1 nodes remain that would care to upgrade).

### CODE REVIEW

File: src/test/txindex_tests.cpp [hunk 1]
```diff
@@ -70,6 +71,8 @@ BOOST_FIXTURE_TEST_CASE(txindex_initial_sync, TestChain100Setup)
     }
 
     txindex.Stop(); // Stop thread before calling destructor
+    LOCK(cs_main);
+    txindex_ptr.reset(nullptr);
 }
```
Comment (MarcoFalke) [inline src/test/txindex_tests.cpp]: Thanks for the txindex test fix. This should be a separate pull request, since the issue is already present (intermittent)

Comment (sipa) [reply src/test/txindex_tests.cpp]: Oh, I was wondering how it could be related...

File: src/validation.cpp [hunk 2]
```diff
@@ -2816,15 +2819,20 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
...
         setBlockIndexCandidates.insert(invalid_walk_tip->pprev);
+
+        // If we abort invalidation after this iteration, make sure
+        // the last disconnected block gets marked failed (rather than
+        // just child of failed)
+        to_mark_failed = invalid_walk_tip;
     }
 
     {
         // Mark the block itself as invalid.
```
Comment (promag) [inline src/validation.cpp]: Commit ebe6eebab3b3a42da70ec99c1f30568ae0abb16f 

Update comment? I mean, can be a different block?

File: src/validation.cpp [hunk 3]
```diff
@@ -2795,6 +2795,11 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
     while (true) {
         if (ShutdownRequested()) break;
 
+        // Make sure the validation callbacks don't overflow
```
Comment (promag) [inline src/validation.cpp]: Commit 1b9518cc7c48750559a39fad4a5056d6a029ee8a

Care to elaborate? Looks like this code could be in a function or something?

File: src/validation.cpp [hunk 4]
```diff
@@ -2811,7 +2812,7 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
         bool ret = DisconnectTip(state, chainparams, &disconnectpool);
         // DisconnectTip will add transactions to disconnectpool; try to add these
         // back to the mempool.
-        UpdateMempoolForReorg(disconnectpool, ret);
+        UpdateMempoolForReorg(disconnectpool, (++disconnected <= 10) && ret);
```
Comment (promag) [inline src/validation.cpp]: Commit 3af503b4a86f3743c8660f15c0a62de0289b7d94

Care to elaborate this optimization here? Looks like it could be:
```cpp
bool ret = DisconnectTip(state, chainparams, ++disconnected <= 10 ? &disconnectpool : nullptr);
...
if (disconnected <= 10) UpdateMempoolForReorg(disconnectpool, ret)
```

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -1583,15 +1583,15 @@ static UniValue invalidateblock(const JSONRPCRequest& request)
     uint256 hash(ParseHashV(request.params[0], "blockhash"));
     CValidationState state;
 
+    CBlockIndex* pblockindex;
     {
         LOCK(cs_main);
         CBlockIndex* pblockindex = LookupBlockIndex(hash);
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: Commit 78122f3d7ed059b00bdaed585:

I think this is invalid cpp and wouldn't compile

Comment (promag) [reply src/rpc/blockchain.cpp]: You mean should just be `pblockindex = LookupBlockIndex(hash);`?

Comment (MarcoFalke) [reply src/rpc/blockchain.cpp]: Yeah, I guess it compiles, but it might do the wrong thing.

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -1586,7 +1586,7 @@ static UniValue invalidateblock(const JSONRPCRequest& request)
     CBlockIndex* pblockindex;
     {
         LOCK(cs_main);
-        CBlockIndex* pblockindex = LookupBlockIndex(hash);
+        pblockindex = LookupBlockIndex(hash);
```
Comment (promag) [inline src/rpc/blockchain.cpp]: This should be in previous commit.

File: src/validation.cpp [hunk 7]
```diff
@@ -4168,37 +4170,105 @@ bool ReplayBlocks(const CChainParams& params, CCoinsView* view) {
...
+            // re-download/re-verify blocks from before segwit actually activated.
+            if (IsWitnessEnabled(chainActive[nHeight - 1], params.GetConsensus()) && !(chainActive[nHeight]->nStatus & BLOCK_OPT_WITNESS)) {
+                break;
+            }
+            nHeight++;
         }
-        nHeight++;
-    }
 
+        // Immediately erase post-segwit blocks without witness not in the main chain.
```
Comment (ryanofsky) [inline src/validation.cpp]: In commit "Release cs_main during RewindBlockIndex operation" (8994e394c0b4e76616c10acdd309d0225eabad9e)

Could you add some rationale in this comment saying why this is done here? Previous behavior of erasing in one place instead of two places seems simpler. Is this an optimization? Is it needed for correctness for some reason now that cs_main is released?

**Review by Sjors — COMMENTED** [review]
Tested 3af503b on macOS 10.14.3 by calling `invalidateblock 0000000000000000001c8018d9cb3b742ef25114f27563e3fc4a1902167f9893` (SegWit activation at height 481,824) from inside the QT console, without setting `-dbcache`.

On master memory usage exceeds 10 GB by block 558,000. It also required a force quit to abort. 

With this PR memory stays well below 1 GB and I'm able to gracefully exit during this. It then marks the most recent block as invalid, so when you restart it doesn't suddenly start syncing to the tip again. For a user upgrading from pre-segwit, would they have to manually restart the process by calling `reconsiderblock [...]`?

Rewinding to SegWit activation took 7 hours on a one year old iMac. At this point it's probably faster to just wipe the chain and sync from scratch, but that's not for this PR. QT sync window won't update during this, but I can live with that (something to solve for the next soft fork...).

File: src/validation.cpp [hunk 8]
```diff
@@ -2789,60 +2789,55 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
-    // are no blocks that meet the "have data and are not invalid per
-    // nStatus" criteria for inclusion in setBlockIndexCandidates).
-
     bool pindex_was_in_chain = false;
-    CBlockIndex *invalid_walk_tip = chainActive.Tip();
 
-    DisconnectedBlockTransactions disconnectpool;
-    while (chainActive.Contains(pindex)) {
+    while (true) {
+        LOCK(cs_main);
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: > Should relax current thread between unlock/lock on each cycle?

I don't understand.

File: src/validation.cpp [hunk 9]
```diff
@@ -2789,9 +2789,12 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
 
 bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& chainparams, CBlockIndex *pindex)
 {
+    CBlockIndex* to_mark_failed = pindex;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: To which scope?

File: src/validation.cpp (hunk 4 above)
Comment (sipa) [reply src/validation.cpp]: No, we still need to call UpdateMempoolForReorg, even if we don't want to add the disconnected transactions back to the mempool.

File: src/validation.cpp [hunk 10]
```diff
@@ -2789,60 +2789,55 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
-            // It's probably hopeless to try to make the mempool consistent
-            // here if DisconnectTip failed, but we can try.
-            UpdateMempoolForReorg(disconnectpool, false);
-            return false;
-        }
-    }
+        DisconnectedBlockTransactions disconnectpool;
+        bool ret = DisconnectTip(state, chainparams, &disconnectpool);
+        // DisconnectTip will add transactions to disconnectpool; try to add these
+        // back to the mempool.
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: I don't know what you're referring to? Is the logic wrong or are you suggested an improvement to the comments?

File: src/validation.cpp [hunk 11]
```diff
@@ -2789,60 +2789,55 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
+        UpdateMempoolForReorg(disconnectpool, ret);
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: Done.

File: src/validation.cpp [hunk 12]
```diff
@@ -2789,60 +2789,55 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
+        if (!ret) return false;
+        assert(invalid_walk_tip->pprev == chainActive.Tip());
 
-    // Now mark the blocks we just disconnected as descendants invalid
-    // (note this may not be all descendants).
-    while (pindex_was_in_chain && invalid_walk_tip != pindex) {
+        // We first disconnect and then mark the blocks as invalid.
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: Improved the comments.

File: src/validation.cpp (hunk 2 above)
Comment (sipa) [reply src/validation.cpp]: Clarified the comment.

File: src/validation.cpp (hunk 3 above)
Comment (sipa) [reply src/validation.cpp]: I've created a small function, and improved the comments ("overflow" wasn't really correct and certainly confusing).

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: FIxed. This was fixed, but in a later commit.

File: src/rpc/blockchain.cpp (hunk 6 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: Done.

File: src/validation.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: I've split that commit into 4 separate ones. I think the flow of changes should be a lot clearer now.

File: src/validation.cpp [hunk 13]
```diff
@@ -4168,37 +4170,105 @@ bool ReplayBlocks(const CChainParams& params, CCoinsView* view) {
...
+    // Mark parent as eligible for main chain again
+    if (index->pprev && index->pprev->IsValid(BLOCK_VALID_TRANSACTIONS) && index->pprev->HaveTxsDownloaded()) {
+        setBlockIndexCandidates.insert(index->pprev);
+    }
+}
+
 bool CChainState::RewindBlockIndex(const CChainParams& params)
 {
-    LOCK(cs_main);
+    CBlockIndex *tip;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/validation.cpp]: I've improved the comments, let me know if it's clear now.

File: src/validation.cpp (hunk 7 above)
Comment (sipa) [reply src/validation.cpp]: I've moved to to a separate step instead of lumping it together with the nHeight calculation loop.

There's no need for it to be done while holding the lock. The reason it's separate is because the disconnect/erase of active blocks should be done simultaneously (so that things get erased even when interrupted, and erasing can't be done before disconnecting), but for non-mainchain blocks we need a separate loop still.

File: src/validation.cpp [hunk 14]
```diff
@@ -4168,6 +4170,42 @@ bool ReplayBlocks(const CChainParams& params, CCoinsView* view) {
...
+    while (ret.first != ret.second) {
+        if (ret.first->second == index) {
+            mapBlocksUnlinked.erase(ret.first++);
+        } else {
+            ++ret.first;
+        }
+    }
+    // Mark parent as eligible for main chain again
+    if (index->pprev && index->pprev->IsValid(BLOCK_VALID_TRANSACTIONS) && index->pprev->HaveTxsDownloaded()) {
+        setBlockIndexCandidates.insert(index->pprev);
```
Comment (Sjors) [inline src/validation.cpp]: IIUC this now happens every block rather than only at the end. I assume there's no other thread potentially messing with `setBlockIndexCandidates` in a way that we care about?

**Review by Sjors — APPROVED** [review]
tACK ac64328

Note that QT doesn't update the block count (etc) as this unwinding happens, although it does update the mempool count. I can live with that for now. 
schermafbeelding 2019-02-25 om 13 35 52

File: src/validation.cpp (hunk 14 above)
Comment (sipa) [reply src/validation.cpp]: This happens with cs_main held.

Comment (Sjors) [reply src/validation.cpp]: Ah I see, the loop over `mapBlockIndex` that calls `EraseBlockData` is entirely inside `LOCK(cs_main)`. So that's indeed the same as before 9d6dcc52c6cb0cdcda220fddccaabb0ffd40068d.

Comment (sipa) [reply src/validation.cpp]: Oh, that too, but that doesn't matter. The entire operation in this function could be called with a different lock per invocation.

File: src/validation.cpp [hunk 15]
```diff
@@ -4208,77 +4208,89 @@ void CChainState::EraseBlockData(CBlockIndex* index)
...
+            // to disk before writing the chainstate, resulting in a failure to continue if interrupted.
+            // Note: If we encounter an insufficiently validated block that
+            // is on chainActive, it must be because we are a pruning node, and
+            // this block or some successor doesn't HAVE_DATA, so we were unable to
+            // rewind all the way.  Blocks remaining on chainActive at this point
+            // must not have their validity reduced.
+            EraseBlockData(tip);
 
+            tip = tip->pprev;
+        }
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/validation.cpp]: Ah, I see this is required in a later commit

**Review by maflcko — APPROVED** [review]
utACK ac64328b96fbe7475d2b12a9a0c2ba4787267204

I left some feedback in earlier commits, but then it solved itself after reading the later commits. Review felt like reading a story that tells itself. Would review again 10/10

**Comment by promag** [comment]
Tested:
 - concurrent `invalidateblock` calls
 - shutdown during `invalidateblock`

After these tests I felt that `reconsiderblock` could return whether the invalidation finished or not. or other meaningful data.

However when testing concurrent `invalidateblock` and `reconsiderblock`, got the following assertion on start:
```
...
2019-02-27T11:53:54Z Checking all blk files are present...
2019-02-27T11:53:54Z Opening LevelDB in /Users/joao/Library/Application Support/Bitcoin/regtest/chainstate
2019-02-27T11:53:54Z Opened LevelDB successfully
2019-02-27T11:53:54Z Using obfuscation key for /Users/joao/Library/Application Support/Bitcoin/regtest/chainstate: 326eeb3c93ca8734
Assertion failed: (!setBlockIndexCandidates.empty()), function PruneBlockIndexCandidates, file validation.cpp, line 2534.
```

To reproduce apply the patch 
```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2840,6 +2840,8 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
         to_mark_failed = invalid_walk_tip;
     }

+    MilliSleep(10000);
+
     {
         // Mark pindex itself as invalid, regardless of whether it was in the main chain or not.
         LOCK(cs_main);
```
Then:
 - `invalidateblock <old block hash>`
 - during the above sleep call `reconsiderblock <previous tip>` 
 - restart `bitcoind`.

**Comment by promag** [comment]
Conceptually, should `InvalidateBlock` interrupt/error if it disconnects the same tip twice? Or should `invalidateblock` and `reconsiderblock` be exclusive? Something to think and probably tackle in a different PR.

**Comment by Sjors** [comment]
I think `invalidateblock` and `reconsiderblock` should be exclusive. `invalidateblock` can be interrupted by exiting `bitcoind` (thanks to this PR), which is good enough UX for something as rare as that. I've only used `invalidateblock` for testing rollback behavior; the only other use case I can think of is a DIY soft fork :-)

**Comment by sipa** [comment]
@promag I think I found what was causing the issue you observed when running reconsiderblock and invalidateblock simultaneously. The final step in the invalidation process would unconditionally mark the last disconnected block as invalid - even if reconsiderblock had made it the active block again at the same time.

Would you mind trying again? Another useful thing to test is to put a sleep inside the first loop in invalidblock (before the cs_main grab), to test the effect of a reconsiderblock in the middle of that loop.

**Comment by promag** [comment]
@sipa I did that, sleeps all over the place, and that case sounded like scratching the chain. I'll repeat the tests.

**Comment by sipa** [comment]
@promag An alternative is only sleeping when "disconnected==1" (which would be in the middle for any invalidate that's longer than 1 deep, but not sleep all the time).

It's possible that the resulting effect is a reconsider and invalidate that start fighting with eachother and disconnect/connect the same blocks over and over again. It's not too hard to detect this scenario inside invalidateblock and bail out, but I'd rather not overload the change here further with such edge cases.

File: src/validation.cpp [hunk 16]
```diff
@@ -2789,64 +2789,74 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
+        to_mark_failed = invalid_walk_tip;
     }
 
-    // Mark the block itself as invalid.
-    pindex->nStatus |= BLOCK_FAILED_VALID;
-    setDirtyBlockIndex.insert(pindex);
-    setBlockIndexCandidates.erase(pindex);
-    m_failed_blocks.insert(pindex);
+    {
+        // Mark pindex (or the last disconnected block) as invalid, regardless of whether it was in the main chain or not.
```
Comment (Sjors) [inline src/validation.cpp]: Nit, move this comment below the interference check, replace "something is interfering" with "some other thread may have interfered before we reestablished the cs_main lock".

File: src/validation.cpp [hunk 17]
```diff
@@ -2789,64 +2789,74 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
-    // add it again.
-    BlockMap::iterator it = mapBlockIndex.begin();
-    while (it != mapBlockIndex.end()) {
-        if (it->second->IsValid(BLOCK_VALID_TRANSACTIONS) && it->second->HaveTxsDownloaded() && !setBlockIndexCandidates.value_comp()(it->second, chainActive.Tip())) {
-            setBlockIndexCandidates.insert(it->second);
+        // The resulting new best tip may not be in setBlockIndexCandidates anymore, so
+        // add it again.
+        BlockMap::iterator it = mapBlockIndex.begin();
+        while (it != mapBlockIndex.end()) {
+            if (it->second->IsValid(BLOCK_VALID_TRANSACTIONS) && it->second->HaveTxsDownloaded() && !setBlockIndexCandidates.value_comp()(it->second, chainActive.Tip())) {
```
Comment (Sjors) [inline src/validation.cpp]: Unchanged behavior, but I'm confused why we only reconsider candidate blocks with `IsValid(BLOCK_VALID_TRANSACTIONS)` && `HaveTxsDownloaded()`, rather than any candidate with just more proof of work that's wasn't marked as invalid. Maybe the answer is: because such block candidates would never have been removed from `setBlockIndexCandidates` in the first place?

**Review by Sjors — APPROVED** [review]
utACK 8d22041. Several improvements in 9bb32eb571a846b66ed3bac493f55cee11a3a1b9 `Release cs_main during InvalidateBlock iterations` since my previous tACK. It's more readable, e.g thanks to the additional variable `to_mark_failed`.

In the event this PR needs further changes (hopefully not) consider splitting that commit; moving `LOCK(cs_main);` into the loop and adding `if (ShutdownRequested()) break;` vs. everything else.

I would also prefer dealing with edge cases with concurrent RPC calls in a later PR. We can point out in the release note that it's currently not 100% safe to do that.

File: src/validation.cpp (hunk 17 above)
Comment (sipa) [reply src/validation.cpp]: That's the definition of `setBlockIndexCandidates`: all blocks with valid downloaded transactions (with optionally the ones that are less cumulative work than the chain tip removed).

Comment (Sjors) [reply src/validation.cpp]: I think I got confused with setBlockIndexHeaderCandidates from #13937.

**Comment by gmaxwell** [comment]
ACK

File: src/validation.cpp [hunk 18]
```diff
@@ -2787,64 +2795,80 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
-    pindex->nStatus |= BLOCK_FAILED_VALID;
-    setDirtyBlockIndex.insert(pindex);
-    setBlockIndexCandidates.erase(pindex);
-    m_failed_blocks.insert(pindex);
+    {
+        // Mark pindex (or the last disconnected block) as invalid, regardless of whether it was in the main chain or not.
+        LOCK(cs_main);
+        if (chainActive.Contains(to_mark_failed)) {
+            // If the to-be-marked invalid block is in the active chain, something is interfering and we can't proceed.
+            return false;
```
Comment (promag) [inline src/validation.cpp]: Just noting that return value is not used.

File: src/validation.cpp [hunk 19]
```diff
@@ -2787,64 +2795,80 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
...
+        }
 
-    // DisconnectTip will add transactions to disconnectpool; try to add these
-    // back to the mempool.
-    UpdateMempoolForReorg(disconnectpool, true);
+        to_mark_failed->nStatus |= BLOCK_FAILED_VALID;
```
Comment (promag) [inline src/validation.cpp]: Couldn't this be inside the loop above, in line 2810?
```cpp
if (ShutdownRequested() || !chainActive.Contains(pindex)) {
    ...
    return
}
```
which would avoid the unlock+lock.

Comment (sipa) [reply src/validation.cpp]: This whole lock scope could indeed be merged into the loop above, but I think it's clearer when it's separate (the loop only deals with disconnecting the main chain if it were to contain pindex, the section afterwards deals with cleanups that are necessary regardless of whether pindex was in the main chain).

Given that we're already locking/unlocking for each disconnected block, I don't think getting rid of one extra lock/unlock matter much.

Comment (promag) [reply src/validation.cpp]: Ok, makes sense.

**Comment by sipa** [comment]
I added an extra commit that solves the situation where if you used `invalidateblock`, had a cache flush before it completed, then had your node crash, and then restarted with `-checkblockindex` would assert with a state corruption.

I've now tested a few more `invalidateblock` scenarios (with `-checkblockindex` and `-checkmempool` enabled), including interrupting it in the middle, killing/crashing it, and running `reconsiderblock` in parallel. They all seem to work fine.

**Comment by gmaxwell** [comment]
re-ACK

**Comment by sipa** [comment]
I think this is ready.
